File: android/src/main/java/com/flightcom/variosound/AudioStreamLooper.java
package com.flightcom.variosound;

/**
 * Continuous real-time variometer tone generator.
 *
 * A free-running sample clock fills fixed-size buffers; every sample the current
 * (smoothed) vertical speed drives the pitch and, when climbing, the beep
 * cadence and duty cycle. See TONE.md for the model. Keep the constants in sync
 * with the iOS version (AudioStreamLooper.swift).
 */
public class AudioStreamLooper {

    public static final int SAMPLE_RATE = 44100;
    private static final double TWO_PI = 2.0 * Math.PI;

    // ---- Tunable tone model (edit to taste; mirror in Swift) ----
    static final double BASE_FREQ = 700.0;   // Hz at 0 m/s
    static final double OCTAVE_PER = 6.0;    // m/s per octave
    static final double FREQ_MIN = 200.0;
    static final double FREQ_MAX = 1600.0;
    static final double CAD_MIN = 1.5;       // beeps/s near 0
    static final double CAD_MAX = 8.0;       // beeps/s at strong climb
    static final double CAD_SLOPE = 1.3;     // beeps/s per m/s
    static final double DUTY_MAX = 0.55;     // on-fraction near 0
    static final double DUTY_MIN = 0.30;     // on-fraction at strong climb
    static final double DUTY_SLOPE = 0.05;   // per m/s
    static final double AMPLITUDE = 0.45;    // 0..1 of full scale
    static final double VZ_SMOOTH = 0.0006;  // per-sample one-pole
    static final double EDGE_S = 0.006;      // attack/release seconds (anti-click)

    // ---- Weak-lift detector ("zérotage") ----
    // A soft, slow, distinct blip for when the air is lifting just enough to
    // offset the glider's own sink (net vz near zero, below the climb
    // threshold). The app turns it on inside that band; see TONE.md.
    static final double WEAK_CAD = 1.2;      // beeps/s (fixed, slow)
    static final double WEAK_DUTY = 0.22;    // short on-fraction
    static final double WEAK_AMP = 0.28;     // softer than the climb tone

    private volatile double targetSpeed = 0.0;
    private volatile boolean weakLift = false;
    private double currentSpeed = 0.0;
    private double phase = 0.0;     // sine phase, radians
    private double beepClock = 0.0; // seconds into the current beep cycle

    public void setSpeed(double v) {
        targetSpeed = v;
    }

    /** Enables the soft weak-lift sound, overriding the climb/sink tone. */
    public void setWeakLift(boolean w) {
        weakLift = w;
    }

    /** Resets the generator for a clean (click-free) start from silence. */
    public void reset() {
        phase = 0.0;
        beepClock = 0.0;
        currentSpeed = targetSpeed;
    }

    public void fill(short[] buffer) {
        for (int i = 0; i < buffer.length; i++) {
            currentSpeed += (targetSpeed - currentSpeed) * VZ_SMOOTH;

            double f = clamp(BASE_FREQ * Math.pow(2.0, currentSpeed / OCTAVE_PER),
                    FREQ_MIN, FREQ_MAX);
            phase += TWO_PI * f / SAMPLE_RATE;
            if (phase >= TWO_PI) phase -= TWO_PI;

            double amp;
            double level = AMPLITUDE;
            if (weakLift) {
                // Weak lift: soft, slow blip, overriding the sign-based tone.
                double period = 1.0 / WEAK_CAD;
                double onTime = period * WEAK_DUTY;
                beepClock += 1.0 / SAMPLE_RATE;
                if (beepClock >= period) beepClock -= period;
                amp = envelope(beepClock, onTime);
                level = WEAK_AMP;
            } else if (currentSpeed < 0.0) {
                // Sink: continuous tone.
                amp = 1.0;
                beepClock = 0.0;
            } else {
                // Climb: beeping, faster and crisper as the climb rate rises.
                double cadence =
                        clamp(CAD_MIN + CAD_SLOPE * currentSpeed, CAD_MIN, CAD_MAX);
                double period = 1.0 / cadence;
                double duty =
                        clamp(DUTY_MAX - DUTY_SLOPE * currentSpeed, DUTY_MIN, DUTY_MAX);
                double onTime = period * duty;

                beepClock += 1.0 / SAMPLE_RATE;
                if (beepClock >= period) beepClock -= period;
                amp = envelope(beepClock, onTime);
            }

            double sample = Math.sin(phase) * amp * level;
            buffer[i] = (short) Math.round(clamp(sample, -1.0, 1.0) * 32767.0);
        }
    }

    /** Trapezoidal beep envelope: attack, sustain, release, then silence. */
    private static double envelope(double t, double onTime) {
        if (t >= onTime) return 0.0;
        double edge = Math.min(EDGE_S, onTime / 2.0);
        if (t < edge) return t / edge;
        if (t > onTime - edge) return (onTime - t) / edge;
        return 1.0;
    }

    private static double clamp(double x, double lo, double hi) {
        return x < lo ? lo : (x > hi ? hi : x);
    }
}

File: ios/Classes/AudioStreamLooper.swift
import Foundation

/// Continuous real-time variometer tone generator (iOS).
///
/// Mirror of the Android `AudioStreamLooper.java`. See TONE.md for the model.
/// Keep the constants below in sync with the Android version.
class AudioStreamLooper {

    static let sampleRate = 44100.0
    private let twoPi = 2.0 * Double.pi

    // ---- Tunable tone model (mirror the Java constants) ----
    private let baseFreq = 700.0   // Hz at 0 m/s
    private let octavePer = 6.0    // m/s per octave
    private let freqMin = 200.0
    private let freqMax = 1600.0
    private let cadMin = 1.5        // beeps/s near 0
    private let cadMax = 8.0        // beeps/s at strong climb
    private let cadSlope = 1.3      // beeps/s per m/s
    private let dutyMax = 0.55      // on-fraction near 0
    private let dutyMin = 0.30      // on-fraction at strong climb
    private let dutySlope = 0.05    // per m/s
    private let amplitude = 0.45    // 0..1 of full scale
    private let vzSmooth = 0.0006   // per-sample one-pole
    private let edgeS = 0.006       // attack/release seconds (anti-click)

    // ---- Weak-lift detector ("zérotage"), mirror the Java constants ----
    private let weakCad = 1.2       // beeps/s (fixed, slow)
    private let weakDuty = 0.22     // short on-fraction
    private let weakAmp = 0.28      // softer than the climb tone

    private var targetSpeed = 0.0
    private var weakLift = false
    private var currentSpeed = 0.0
    private var phase = 0.0     // sine phase, radians
    private var beepClock = 0.0 // seconds into the current beep cycle

    func setSpeed(_ v: Double) {
        targetSpeed = v
    }

    /// Enables the soft weak-lift sound, overriding the climb/sink tone.
    func setWeakLift(_ w: Bool) {
        weakLift = w
    }

    /// Resets the generator for a clean (click-free) start from silence.
    func reset() {
        phase = 0.0
        beepClock = 0.0
        currentSpeed = targetSpeed
    }

    func fill(_ buffer: UnsafeMutablePointer<Float>, _ count: Int) {
        for i in 0..<count {
            currentSpeed += (targetSpeed - currentSpeed) * vzSmooth

            let f = clamp(baseFreq * pow(2.0, currentSpeed / octavePer),
                          freqMin, freqMax)
            phase += twoPi * f / AudioStreamLooper.sampleRate
            if phase >= twoPi { phase -= twoPi }

            var amp: Double
            var level = amplitude
            if weakLift {
                // Weak lift: soft, slow blip, overriding the sign-based tone.
                let period = 1.0 / weakCad
                let onTime = period * weakDuty
                beepClock += 1.0 / AudioStreamLooper.sampleRate
                if beepClock >= period { beepClock -= period }
                amp = envelope(beepClock, onTime)
                level = weakAmp
            } else if currentSpeed < 0.0 {
                // Sink: continuous tone.
                amp = 1.0
                beepClock = 0.0
            } else {
                // Climb: beeping, faster and crisper as the climb rate rises.
                let cadence = clamp(cadMin + cadSlope * currentSpeed, cadMin, cadMax)
                let period = 1.0 / cadence
                let duty = clamp(dutyMax - dutySlope * currentSpeed, dutyMin, dutyMax)
                let onTime = period * duty

                beepClock += 1.0 / AudioStreamLooper.sampleRate
                if beepClock >= period { beepClock -= period }
                amp = envelope(beepClock, onTime)
            }

            let sample = sin(phase) * amp * level
            buffer[i] = Float(clamp(sample, -1.0, 1.0))
        }
    }

    /// Trapezoidal beep envelope: attack, sustain, release, then silence.
    private func envelope(_ t: Double, _ onTime: Double) -> Double {
        if t >= onTime { return 0.0 }
        let edge = min(edgeS, onTime / 2.0)
        if t < edge { return t / edge }
        if t > onTime - edge { return (onTime - t) / edge }
        return 1.0
    }

    private func clamp(_ x: Double, _ lo: Double, _ hi: Double) -> Double {
        return x < lo ? lo : (x > hi ? hi : x)
    }
}
